From 8a5f67b96e57c9e41863f22100182c89684bb088 Mon Sep 17 00:00:00 2001
From: Christian Hergert <chergert@redhat.com>
Date: Tue, 19 Nov 2024 18:03:40 -0800
Subject: [PATCH 2/2] osc: implement 9;4; OSC for progress tracking

Windows Terminal implemented ConEmu's OSC 9;4; progress tracking OSC to
allow terminals to display progress information in the tab, window, and
taskbar icon.

Systemd 247 will support exporting this progress tracking for most
internal commands which have progress.

As we never implemented iTerm2's GROWL OSC, this doesn't break existing
code using the OSC and is specific to checking for 9;4;.

This could be extended to support the various states, but that is left
as future exercise as it is not implemented in systemd and thus has less
immediate value.
---
 src/parser-osc.hh  |  3 ++-
 src/vteinternal.hh |  3 +++
 src/vteseq.cc      | 60 +++++++++++++++++++++++++++++++++++++++++++++-
 3 files changed, 64 insertions(+), 2 deletions(-)

diff --git a/src/parser-osc.hh b/src/parser-osc.hh
index e6391a21..9729806c 100644
--- a/src/parser-osc.hh
+++ b/src/parser-osc.hh
@@ -21,11 +21,12 @@
 
 /* _VTE_OSC(DTTERM_CWD, 3) Conflicts with XTERM_SET_XPROPERTY */
 
+_VTE_OSC(CONEMU_EXTENSION, 9)
+
 _VTE_OSC(EMACS_51, 51)
 
 _VTE_OSC(FOOT_FLASH, 555)
 
-_VTE_OSC(ITERM2_GROWL, 9)
 _VTE_OSC(ITERM2_SHELL_INTEGRATION, 133)
 _VTE_OSC(ITERM2_1337, 1337)
 
diff --git a/src/vteinternal.hh b/src/vteinternal.hh
index 051e78c8..dcde74df 100644
--- a/src/vteinternal.hh
+++ b/src/vteinternal.hh
@@ -1801,6 +1801,9 @@ public:
         void urxvt_extension(vte::parser::Sequence const& seq,
                              vte::parser::StringTokeniser::const_iterator& token,
                              vte::parser::StringTokeniser::const_iterator const& endtoken) noexcept;
+        void conemu_extension(vte::parser::Sequence const& seq,
+                              vte::parser::StringTokeniser::const_iterator& token,
+                              vte::parser::StringTokeniser::const_iterator const& endtoken) noexcept;
 
         // helpers
 
diff --git a/src/vteseq.cc b/src/vteseq.cc
index b41c99b1..ac6b5db3 100644
--- a/src/vteseq.cc
+++ b/src/vteseq.cc
@@ -2192,6 +2192,61 @@ catch (...)
         // nothing to do here
 }
 
+void
+Terminal::conemu_extension(vte::parser::Sequence const& seq,
+                           vte::parser::StringTokeniser::const_iterator& token,
+                           vte::parser::StringTokeniser::const_iterator const& endtoken) noexcept
+try
+{
+        if (token == endtoken)
+                return;
+
+        auto maybe_set_termprop = [&](int prop,
+                                      auto&& value) -> void {
+                if (auto const info = get_termprop_info(prop)) {
+                        m_termprops_dirty.at(info->id()) = true;
+                        m_termprop_values.at(info->id()) = std::move(value);
+                        m_pending_changes |= vte::to_integral(PendingChanges::TERMPROPS);
+                }
+        };
+
+        auto maybe_reset_termprop = [&](int prop) -> void {
+                if (auto const info = get_termprop_info(prop)) {
+                        m_termprops_dirty.at(info->id()) = true;
+                        m_termprop_values.at(info->id()) = {};
+                        m_pending_changes |= vte::to_integral(PendingChanges::TERMPROPS);
+                }
+        };
+
+        auto const cmd = *token;
+        if (++token == endtoken)
+                return;
+
+        if (cmd == "4") {
+                auto const st = *token;
+                if (++token == endtoken)
+                        return;
+
+                if (st == "0") {
+                        maybe_reset_termprop(VTE_PROPERTY_ID_PROGRESS);
+
+                } else if (st == "1") {
+                        int pr;
+
+                        if (!token.number(pr) || pr < 0 || pr > 100) {
+                                return;
+                        }
+
+                        maybe_set_termprop(VTE_PROPERTY_ID_PROGRESS, uint64_t(pr));
+
+                }
+        }
+}
+catch (...)
+{
+        // nothing to do here
+}
+
 // collect_rect:
 // @seq:
 // @idx:
@@ -7555,6 +7610,10 @@ Terminal::OSC(vte::parser::Sequence const& seq)
                 urxvt_extension(seq, it, cend);
                 break;
 
+        case VTE_OSC_CONEMU_EXTENSION:
+                conemu_extension(seq, it, cend);
+                break;
+
         case VTE_OSC_XTERM_SET_ICON_TITLE:
         case VTE_OSC_XTERM_SET_XPROPERTY:
         case VTE_OSC_XTERM_SET_COLOR_MOUSE_CURSOR_FG:
@@ -7574,7 +7633,6 @@ Terminal::OSC(vte::parser::Sequence const& seq)
         case VTE_OSC_XTERM_RESET_COLOR_TEK_CURSOR:
         case VTE_OSC_EMACS_51:
         case VTE_OSC_ITERM2_1337:
-        case VTE_OSC_ITERM2_GROWL:
         case VTE_OSC_KONSOLE_30:
         case VTE_OSC_KONSOLE_31:
         case VTE_OSC_RLOGIN_SET_KANJI_MODE:
-- 
2.47.0

